### PR TITLE
Builtin Log Handler

### PR DESCRIPTION
**Pull request by jnealtowns**
Reviewer: @kenchiang  @rlane 

This log handler supports the AIM_SYSLOG_* and AIM_LOG_* macros.
It also supports logging to stdout, stderr, and an optional debug log file.

This code was ported from private code initially implemented by @kenchiang.

### CODE REVIEW

File: modules/AIM/module/inc/AIM/aim_log_handler.h [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+/**
+ * @brief Basic initialization for console and daemonized clients.
+ * @param ident The syslog ident to use (optional)
+ * @param debug_log_file  The name of the debug log file (optional)
+ * @param max_debug_size   Maximum debug log size.
+ * @param max_debug_count  Maximum number of rotated debug logs.
+ *
+ * @note This is designed to be a simple and generic initialization
+ * for both daemonized and console-based clients. It will initialize
```
Comment (kenchiang) [inline modules/AIM/module/inc/AIM/aim_log_handler.h]: comment seems truncated

File: modules/AIM/module/inc/AIM/aim_log_handler.h [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+ */
+int aim_log_handler_basic_init_all(const char* ident,
+                                    const char* debug_log_file,
+                                    int max_debug_log_size,
+                                    int max_debug_logs);
```
Comment (kenchiang) [inline modules/AIM/module/inc/AIM/aim_log_handler.h]: minor: alignment

File: modules/AIM/module/src/aim_log_handler.c [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+static void
+rotate_debug_log__(aim_log_handler_t handler)
+{
+    struct stat fp_log_stat;
+    if (stat(handler->config.debug_log_name, &fp_log_stat) != -1) {
+        if (fp_log_stat.st_size >= handler->config.max_debug_log_size) {
+
+            int debug_log_name_len = strlen(handler->config.debug_log_name);
+            char* src = aim_malloc(debug_log_name_len + 16);
+            char* dst = aim_malloc(debug_log_name_len + 16);
```
Comment (rlane) [inline modules/AIM/module/src/aim_log_handler.c]: Memory leak.

File: modules/AIM/module/inc/AIM/aim_config.h [hunk 4]
```diff
@@ -209,7 +209,7 @@ struct aim_pvs_s;
 
 
 #ifndef AIM_CONFIG_INCLUDE_DAEMONIZE
-#define AIM_CONFIG_INCLUDE_DAEMONIZE 0
+#define AIM_CONFIG_INCLUDE_DAEMONIZE 1
```
Comment (rlane) [inline modules/AIM/module/inc/AIM/aim_config.h]: Is this change meant to be in this pull request?

File: modules/AIM/module/inc/AIM/aim_log.h [hunk 5]
```diff
@@ -234,9 +234,7 @@ typedef struct aim_log_s {
  * Reasonable default option settings.
  */
 #ifndef AIM_LOG_OPTIONS_DEFAULT
-#define AIM_LOG_OPTIONS_DEFAULT                  \
-    ( AIM_LOG_OPTION_BIT_ENABLE +                \
-      AIM_LOG_OPTION_BIT_TIMESTAMP )
+#define AIM_LOG_OPTIONS_DEFAULT AIM_LOG_OPTION_BIT_ENABLE
```
Comment (rlane) [inline modules/AIM/module/inc/AIM/aim_log.h]: Don't change the default, it will break existing applications.

**Comment by rlane** [comment]
Finished review.

File: modules/AIM/module/inc/AIM/aim_log.h (hunk 5 above)
Comment (jnealtowns) [inline modules/AIM/module/inc/AIM/aim_log.h]: Unintentional. Reverted.

File: modules/AIM/module/inc/AIM/aim_config.h (hunk 4 above)
Comment (jnealtowns) [inline modules/AIM/module/inc/AIM/aim_config.h]: Not really, but its already changed to 1 from 0 in master as a result of a previous pull request. I don't know why it thinks its a new change.

**Comment by jnealtowns** [comment]
SELF-ACK

**Comment by bsn-abat** [comment]
ABAT: START_MERGE:  Log file at http://vbs24.eng.bigswitch.com/abat/2016.03.01.2013-m.infra.master/abat.log

**Comment by bsn-abat** [comment]
ABAT: ACCEPT:  Successfully merged

In case you want to see the build log, check out :
Log file at http://vbs24.eng.bigswitch.com/abat/2016.03.01.2013-m.infra.master/abat.log
/cc @rlane @jnealtowns

**Pull request merged**